aptos-move/framework/aptos-stdlib/sources/big_vector.move

```diff
@@ -0,0 +1,373 @@
+module aptos_std::big_vector {
+    use std::error;
+    use std::vector;
+    use aptos_std::table_with_length::{Self, TableWithLength};
+
+    friend aptos_std::smart_vector;
+
+    /// Vector index is out of bounds
+    const EINDEX_OUT_OF_BOUNDS: u64 = 1;
+    /// Cannot destroy a non-empty vector
+    const EVECTOR_NOT_EMPTY: u64 = 2;
+    /// bucket_size cannot be 0
+    const EZERO_BUCKET_SIZE: u64 = 3;
+
+    /// A scalable vector implementation based on tables where elements are grouped into buckets.
+    /// Each bucket has a capacity of `bucket_size` elements.
+    struct BigVector<T> has store {
+        buckets: TableWithLength<u64, vector<T>>,
+        end_index: u64,
+        num_buckets: u64,
+        bucket_size: u64
+    }
+
+    /// Regular Vector API
+
+    /// Create an empty vector.
+    public(friend) fun empty<T: store>(bucket_size: u64): BigVector<T> {
+        assert!(bucket_size > 0, error::invalid_argument(EZERO_BUCKET_SIZE));
+        BigVector {
+            buckets: table_with_length::new(),
+            end_index: 0,
+            num_buckets: 0,
+            bucket_size,
+        }
+    }
+
+    /// Create a vector of length 1 containing the passed in element.
+    public(friend) fun singleton<T: store>(e: T, bucket_size: u64): BigVector<T> {
```

Comment:
I made these two friend only to make this struct private for friend modules. Is this the right way to do that? @movekevin

Reply:
Why do these need to be friend? I.e. why can't these be open for anyone to use?